deployment: deploy statics to S3

fabiosantoscode opened the pull request:
So here's what I've been working on this week.

It's not ready to merge, because the script that destroys the PR isn't working for some reason. I can't debug it since there are no logs in the destroy script, so on monday, I'll set up an ngrok and POST stuff to my laptop so I can read errors.

## Strategy

I've replaced the build command only on heroku, such that it:

1. Downloads the public/ folder from S3 (if it's a cold PR it gets the production folder since there are no secrets there)
2. Runs gatsby build
3. Uploads the public/ folder to S3.

The production server then merely proxies to S3. It doesn't work for your local public/ folder.

## Timings

Production build: 3min
First PR build: 4min
Subsequent builds of the same PR: 3min

Downloading and uploading from S3 takes very little time. I was able to increase the request concurrency to 50, so the initial download only takes 10 seconds. Uploading to an already-hot S3 bucket takes around 3 seconds since most files will already be there.

## Drawbacks

In AWS, I've created two IAM users, one for production (which has write access to the production prefix of my bucket), and another for PRs (which has access to $bucket/pulls/*). This means that if someone can leak the tokens in their build (by changing the build command) they can mess with other people's PRs. Therefore, we shouldn't enable automatic PR builds.

shcheklein commented on scripts/deploy-with-s3.js:
```diff
@@ -0,0 +1,107 @@
+
+async function main() {
+  const emptyPrefix = await prefixIsEmpty(s3Prefix);
+  // TODO if the prefix is empty, we can have S3 copy everything
```
let's not put is a TODO in code - let's create a ticker for this

fabiosantoscode replied:
Sure, sorry.

shcheklein commented on scripts/deploy-with-s3.js:
```diff
@@ -0,0 +1,107 @@
+    // Clear it and try again.
+    await remove(cacheDir);
+    await remove(publicDir);
+    run('yarn build');
```
minor: is there a lib that wraps a call into retry?

fabiosantoscode replied:
There's a bunch but in this case we need an action between the retries.

shcheklein commented on scripts/deploy-with-s3.js:
```diff
@@ -0,0 +1,107 @@
+    console.error('------------------------\n\n');
+    console.error(buildError);
+    console.error('\nAssuming bad cache and retrying:\n');
+    // Sometimes gatsby build fails because of bad cache.
```
minor: probably, move comment up

shcheklein commented on scripts/destroy-pr.js:
```diff
@@ -0,0 +1,23 @@
+const { s3Prefix, s3Bucket, s3Client } = require('./s3-utils');
+
+async function main() {
+  console.log(`Deleting s3://${s3Bucket}/${s3Prefix}...`);
```
we can actually setup a lifecycle policy to destroy stuff after a month just to be safe ... may be even remove this code

fabiosantoscode replied:
I can't get logs off this script to understand why it's failing or even whether it's running so I thought of doing just that over the weekend.

shcheklein commented:
> This means that if someone can leak the tokens in their build (by changing the build command) they can mess with other people's PRs. Therefore, we shouldn't enable automatic PR builds.

could you elaborate on this please, not sure I got the idea.

shcheklein commented on app.json:
```diff
@@ -12,6 +12,8 @@
   "name": "dvc-blog-production",
-  "scripts": {},
+  "scripts": {
+    "pr-predestroy": "./scripts/destroy-pr.js"
```
see my comment? may be we can simplify with lifecycle policies. It also beneficial since PR tend to expire after 5 days and it's actually better to reuse bucket if we get it up again.

shcheklein commented on scripts/deploy-with-s3.js:
```diff
@@ -0,0 +1,107 @@
+    console.timeEnd('download from s3');
+  } catch (e) {
+    console.error('Error downloading initial data');
+    console.error(e);
```
is it possible to do `console.error(msg, e`)?

fabiosantoscode replied:
I'll check

fabiosantoscode replied:
Can do

shcheklein commented on scripts/deploy-with-s3.js:
```diff
@@ -0,0 +1,107 @@
+
+  // First build of a PR is slow because it can't reuse cache.
+  // But we can download from prod to warm cache up.
+  const cacheWarmPrefix = emptyPrefix ? 'prod' : s3Prefix;
```
ideally, we should not be using the same `prod` prefix, we should be generating a new one for prod every time to make the operation atomic

fabiosantoscode replied:
Heroku doesn't generate a different environment variable for each production deployment. Plus, it doesn't provide a script to destroy production environments.

The deployment takes around 3 seconds to upload (given a low amount of files) so it's pretty atomic as it is. Plus, the cache shields us from inconsistency as long as it's warm, and gets cleared in one go.

shcheklein commented on .gitignore:
```diff
@@ -7,3 +7,4 @@ tmp/*
 .DS_Store
 .idea
 .vscode
+404.html
```
why do we need to ignore it?

fabiosantoscode replied:
If you run the deploy script locally this file goes to the root and shouldn't be committed since it's generated.

shcheklein commented on src/server/server.js:
```diff
@@ -8,38 +10,77 @@
+const { PORT = 9000, NODE_ENV } = process.env;
+
+const cacheControl = 'public, max-age=0, s-maxage=999999';
+const wreckOptions = {
```
do we need it as a global const?

fabiosantoscode replied:
Not really. I'll place this right in the call.

shcheklein commented on src/server/server.js:
```diff
@@ -8,38 +10,77 @@
+
+let notFoundPage;
+try {
+  // in production there's no public folder
```
still not sure what's going on with 404

fabiosantoscode replied:
This is just to avoid the extra S3 request to fetch the 404 page, which we know doesn't change until the deployment dies.

We don't deploy the public folder so 404.html is moved to the root.

shcheklein commented on src/server/server.js:
```diff
@@ -8,38 +10,77 @@
   /* tslint:disable-next-line:no-console */
-  console.log(`Listening on http://0.0.0.0:${port}/`);
+  console.log(`Listening on http://0.0.0.0:${PORT}/`);
+  console.log(`Proxying to ${s3Url}`);
```
what about local run?

shcheklein replied:
does it still deploy to S3?

shcheklein replied:
or we don't use express middleware locally at all? when we do `yarn develop`

fabiosantoscode replied:
yarn develop is still powered by Gatsby develop.

This is the production server, and you can run it locally with the correct environment variables. And it won't serve the public folder, instead it proxies to S3 just like production. You can also deploy locally if you need to but you need the right AWS token to do so.

I haven't really documented what variables this needs to run, so I'll do that.

shcheklein replied:
👍

shcheklein commented on src/server/server.js:
```diff
@@ -8,38 +10,77 @@
+    await serveFile(pathname, res);
+  } catch (e) {
+    try {
+      const indexFile = (pathname + '/index.html').replace(/\/+/, '/');
```
why do we need this replace?

fabiosantoscode replied:
The pathname might have a trailing slash in it.

fabiosantoscode commented:
@shcheklein I believe I've addressed the comments specific to code, and am now working on the lifecycle policy for the bucket.

>  Didn't get the issue with IAM.

Basically, a bad actor can send a PR which runs `env` during the build, and reads the heroku CI logs therefore gaining access to the AWS token. This can only happen if we enable automatic PRs or run preview environments carelessly.

However this cannot compromise production (separate tokens, separate ACLs), just other PRs.

shcheklein commented on src/server/server.js:
```diff
@@ -1,45 +1,94 @@
+
 /*
- * Production server.
+ * Production server. Proxies to S3 depending on HEROKU_APP_NAME (see scripts/deploy-with-s3.js)
```
minor: looks >80 long to mee

fabiosantoscode commented:
Preview environment for this PR, served from S3, right here: https://dvc-blog-master-dmu0xfokikp14f.herokuapp.com/
